consensus/clique/clique.go

```diff
@@ -504,27 +504,36 @@ func (c *Clique) Prepare(chain consensus.ChainReader, header *types.Header) erro
 	header.Nonce = types.BlockNonce{}
 
 	number := header.Number.Uint64()
+	// Assemble the voting snapshot
+	snap, err := c.snapshot(chain, number-1, header.ParentHash, nil)
+	if err != nil {
+		return err
+	}
 	if number%c.config.Epoch != 0 {
+		// Get valid votes
 		c.lock.RLock()
-		if len(c.proposals) > 0 {
-			addresses := make([]common.Address, 0, len(c.proposals))
-			for address := range c.proposals {
+		var addresses []common.Address
+		var authorizes []bool
```

Comment:
This `authorizes` slice complicates the code a bit. You could drop it entirely, and instead of `authorizes[index]` below in line 528 you could leave the original `c.proposals[header.Coinbase]`. Imho that would make the code cleaner and less error prone.